# Analisis del equilibrio

## Importación de librerías y lectura de datos

In [22]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [26]:
df_x = pd.read_csv("../resultados/dinamica/velocidad_x.txt", delim_whitespace=True, header=None)
df_y = pd.read_csv("../resultados/dinamica/velocidad_y.txt", delim_whitespace=True, header=None)
df_z = pd.read_csv("../resultados/dinamica/velocidad_z.txt", delim_whitespace=True, header=None)

Datos = pd.read_csv("../resultados/dinamica/gname2.txt",skiprows=1, delim_whitespace=True, header=None, names=['t','Ec','Ep','E_tot','W','T','P_corr','F2'])
t = pd.read_csv("../resultados/dinamica/gname2.txt",skiprows=1, delim_whitespace=True, header=None, names=['t','Ec','Ep','E_tot','W','T','P_corr','F2']).t.to_numpy()



df_x.describe()

/var/folders/15/rn_wb7ms2yq2yn7zj6rqkq880000gn/T/ipykernel_3972/1296389926.py:1: FutureWarning: The 'delim_whitespace' keyword in pd.read_csv is deprecated and will be removed in a future version. Use ``sep='\s+'`` instead
  df_x = pd.read_csv("../resultados/dinamica/velocidad_x.txt", delim_whitespace=True, header=None)
/var/folders/15/rn_wb7ms2yq2yn7zj6rqkq880000gn/T/ipykernel_3972/1296389926.py:2: FutureWarning: The 'delim_whitespace' keyword in pd.read_csv is deprecated and will be removed in a future version. Use ``sep='\s+'`` instead
  df_y = pd.read_csv("../resultados/dinamica/velocidad_y.txt", delim_whitespace=True, header=None)
/var/folders/15/rn_wb7ms2yq2yn7zj6rqkq880000gn/T/ipykernel_3972/1296389926.py:3: FutureWarning: The 'delim_whitespace' keyword in pd.read_csv is deprecated and will be removed in a future version. Use ``sep='\s+'`` instead
  df_z = pd.read_csv("../resultados/dinamica/velocidad_z.txt", delim_whitespace=True, header=None)
/var/folders/15/rn_wb7ms2yq2yn7zj6

0           1           2           3           4           5    \
count  500.000000  500.000000  500.000000  500.000000  500.000000  500.000000   
mean     0.269320   -0.127096   -0.431061    0.088092    0.022665    0.189134   
std      1.276415    1.429853    1.239342    1.121372    0.916396    0.855558   
min     -3.442022   -3.335296   -3.055128   -3.351409   -1.668998   -1.803684   
25%     -0.454549   -1.269003   -1.216148   -0.643535   -0.638749   -0.315506   
50%      0.302178   -0.008320   -0.288623    0.003807    0.076378    0.272986   
75%      1.451527    0.951423    0.528457    0.893266    0.678445    0.846028   
max      2.109284    2.592654    2.037268    2.445749    1.855424    1.830911   

              6           7           8           9    ...         490  \
count  500.000000  500.000000  500.000000  500.000000  ...  500.000000   
mean    -0.337620    0.176112   -0.160773   -0.577318  ...    0.159913   
std      1.151500    1.108179    0.935754    1.265076  ...    1.508145   
min     -3.564496   -1.933489   -1.901019   -2.722160  ...   -3.342081   
25%     -1.052561   -0.708093   -0.740551   -1.819378  ...   -0.918104   
50%     -0.099454    0.160149   -0.256177   -0.259714  ...    0.118973   
75%      0.506342    1.032792    0.174736    0.248875  ...    1.442732   
max      1.650593    3.023276    2.644614    2.199851  ...    3.054464   

              491         492         493         494         495         496  \
count  500.000000  500.000000  500.000000  500.000000  500.000000  500.000000   
mean    -0.193354   -0.040535    0.024096   -0.344931   -0.152053   -0.528461   
std      0.825465    1.237660    1.150842    1.106702    1.356028    0.949409   
min     -2.736501   -2.194421   -2.627592   -2.998855   -3.691939   -2.497056   
25%     -0.758936   -1.165724   -1.090009   -1.195678   -0.812521   -1.160026   
50%     -0.127395   -0.348719    0.034806   -0.212596   -0.108044   -0.578383   
75%      0.356638    0.989843    0.836196    0.516400    0.800063    0.209198   
max      1.522626    3.009515    2.317379    2.331283    2.398712    1.512697   

              497         498         499  
count  500.000000  500.000000  500.000000  
mean    -0.522814   -0.108904    0.135523  
std      1.215310    0.727186    0.766910  
min     -3.296692   -1.699571   -1.397501  
25%     -1.132052   -0.572691   -0.391559  
50%     -0.505416   -0.036047    0.099867  
75%      0.390855    0.305136    0.591726  
max      2.457799    2.015852    2.242265  

[8 rows x 500 columns]

### Formato de datos importados

- En las columnas tenemos las particulas de la caja etiquetadas del 0 al 499
- En las filas tenemos instantes sucesivos de tiempo grabados de nuestra simulacion.

In [27]:
df_x.head()

0         1         2         3         4         5         6    \
0  1.347958  0.785519  0.156655 -0.363839  0.354594 -0.083723 -1.024259   
1  1.379724  0.711568  0.068022 -0.356285  0.359387 -0.315906 -1.104741   
2  1.413691  0.631462 -0.091426 -0.349096  0.368068 -0.508772 -1.150660   
3  1.449855  0.547083 -0.325638 -0.342362  0.373095 -0.654217 -1.174325   
4  1.488146  0.460642 -0.615457 -0.336151  0.366230 -0.753281 -1.184756   

        7         8         9    ...       490       491       492       493  \
0 -0.303688  2.477313  0.485794  ... -0.961377 -0.028427  1.871403  2.068640   
1 -0.183217  2.546416  0.450819  ... -0.975145 -0.046258  1.876519  2.024856   
2 -0.002359  2.597422  0.404519  ... -0.980005 -0.047553  1.872005  1.981554   
3  0.224731  2.630290  0.349328  ... -0.973329 -0.023300  1.865353  1.938581   
4  0.488922  2.644614  0.288815  ... -0.952106  0.028329  1.864051  1.894286   

        494       495       496       497       498       499  
0  0.208181 -1.039646  1.495183 -0.998233 -0.570794  0.483009  
1  0.142911 -1.181704  1.369217 -0.998654 -0.369613  0.468339  
2  0.044930 -1.316452  1.231946 -0.990114 -0.187949  0.455757  
3 -0.109352 -1.433872  1.092505 -0.967899 -0.034554  0.446998  
4 -0.334101 -1.526896  0.957570 -0.925776  0.087323  0.443604  

[5 rows x 500 columns]

In [28]:
Datos.head()

t        Ec        Ep     E_tot         W         T      P_md    P_corr  \
0  0.01  1035.798 -1700.832 -665.0342 -1143.778  1.381064  0.275764  611.2330   
1  0.02  1034.588 -1699.631 -665.0429 -1122.042  1.379451  0.282203  630.6700   
2  0.03  1037.483 -1702.533 -665.0501 -1158.830  1.383311  0.271870  595.8498   
3  0.04  1042.828 -1707.877 -665.0483 -1229.975  1.390438  0.251719  556.1786   
4  0.05  1048.210 -1713.257 -665.0470 -1299.122  1.397613  0.232257  532.4419   

   F2  
0 NaN  
1 NaN  
2 NaN  
3 NaN  
4 NaN

Para las distribuciones de velocidad interesan solamente la muestra de velocidades. Independientemente del tiempo. Por lo tanto aplanaremos nuestras DF en un array con todas las velocidades

In [ ]:
df_x.values.flatten() # Devuelve un array uniendo fila a fila del DF

## Preparo distribución de velocidades

In [ ]:
vx=df_x.values.flatten()
vy=df_y.values.flatten()
vz=df_z.values.flatten()

## Calculo temperaturas
En unidades reducidas $<v_x^2>=T$

In [ ]:
vx2=vx*vx
vy2=vy*vy
vz2=vz*vz

Tx=vx2.mean();#Tx_var=vx*(vx.std())
Ty=vy2.mean();#Ty_var=vy2.var()
Tz=vz2.mean();#Tz_var=vz2.var()

In [ ]:
print(f"Tx={Tx:.4f}()\nTy={Ty:.4f}()\nTz={Tz:.4f}()")

# Graficas de las distribuciones de velocidad

### Función distribución de Maxwell

In [ ]:
def Maxwell(v,T):
    return (1/np.sqrt(2*np.pi*T))*np.exp(-(v*v)/(2*T))

#x = np.linspace(vx.min(), vx.max(), 200)
#plt.plot(x,Maxwell(x,1))

In [ ]:
#from scipy.stats import norm

plt.hist(vx, bins=60, density=True, alpha=0.6, color='skyblue', label='Simulación')


x = np.linspace(vx.min(), vx.max(), 200)
y = Maxwell(x,Tx)
#plt.plot(x, norm.pdf(x, 0, np.sqrt(Tx)), 'r-', lw=2, label='Teoría (Maxwell)')
plt.plot(x,y,'r-', lw=2, label="Distribución de Maxwell")


plt.xlabel("Velocidades")
plt.ylabel("Probabilidades")
plt.title("Distribución de velocidad en eje x")
plt.grid(True)
plt.minorticks_on()
plt.legend(loc='upper right', framealpha=0.3)
plt.tight_layout()
plt.savefig("Vx.png") 
#plt.show()

In [ ]:
plt.hist(vy, bins=70, density=True, alpha=0.6, color='green', label='Simulación')

x = np.linspace(vy.min(), vy.max(), 200)
y = Maxwell(x,Ty)

plt.plot(x,y,'r-', lw=2, label="Distribución de Maxwell")


plt.xlabel("Velocidades")
plt.ylabel("Probabilidades")
plt.title("Distribución de velocidad en eje y")
plt.grid(True)
plt.minorticks_on()
plt.legend(loc='upper right', framealpha=0.3)
plt.tight_layout()
plt.savefig("Vy.png") 
#plt.show()

In [ ]:
plt.hist(vz, bins=60, density=True, alpha=0.6, color='orange', label="Distribución de velocidades")

x = np.linspace(vz.min(), vz.max(), 200)
y = Maxwell(x,Tz)

plt.plot(x,y,'r-', lw=2, label="Distribución de Maxwell")


plt.xlabel("Velocidades")
plt.ylabel("Probabilidades")
plt.title("Distribución de velocidad en eje z")
plt.grid(True)
plt.minorticks_on()
plt.legend(loc='upper right', framealpha=0.3)
plt.tight_layout()
plt.savefig("Vz.png") 
#plt.show()

# La función H de Boltzman

$$H(t) = \int f(\mathbf{v},t)\,\ln f(\mathbf{v},t)\,d\mathbf{v}$$ con $\int f(\mathbf{v},t)\,d\mathbf{v} = 1$

### El teorema H de Boltzman
El teorema H de Boltzman establece que $\frac{dH}{dt} \le 0$  y que H alcanza un mínimo constante cuando el sistema llega al equilibrio, momento en el cual $f(\mathbf{v})$ tiene la forma de Maxwell–Boltzmann.


In [ ]:
df_x

index=np.arange(0,len(df_x)) // 50

if "tiempo" not in  df_x.columns:
    df_x.insert(0, "tiempo", t)
    
df_x["index"]=index
H=df_x.set_index("index").groupby("index").apply(lambda df: pd.Series({"tiempo_medio":df.tiempo.mean(),"Distribucion":df.iloc[:,1:].to_numpy().flatten()}))


In [ ]:
f, edges = np.histogram(vx, bins=100, density=True)
v_mid = 0.5*(edges[1:] + edges[:-1]) 


#Tengo que hacer varios f(v) por ejemplo agrupando los vx cada 200 y asignandoles el valor de su tiempo medio 
# La grafica no tiene que ser perfecta, solo hay que ver que deja de caer la H
# Entonces cada 200 grabaciones hago un f(v) y le asigno t=t_medio
# Integro los f(v) y saco una H para cada 200 pasos.
# Hago grafica de H

def simpson_discreto(x, y):
    n = len(x)
    if n % 2 == 0:
        n -= 1  # asegurar impar para Simpson
        x, y = x[:n], y[:n]
    h = (x[-1]-x[0])/(n-1)
    S = y[0] + y[-1] + 4*np.sum(y[1:-1:2]) + 2*np.sum(y[2:-2:2])
    return h*S/3

def H_boltzmann(v):
    # construimos histograma normalizado f(v)
    f, bins = np.histogram(v, bins=100, density=True)
    v_medios = 0.5 * (bins[1:] + bins[:-1])
    # evitamos log(0)
    f_nonzero = f[f > 0]
    v_nonzero = v_medios[f > 0]
    # calculamos H = ∫ f ln f dv
    return simpson_discreto(v_nonzero, f_nonzero * np.log(f_nonzero))

H["H_Boltzman"]=H.Distribucion.map(H_boltzmann)



H


In [ ]:
plt.plot(H.tiempo_medio,H.H_Boltzman, "r-")

print(f"<H>={H.H_Boltzman.mean():.4f} ± {H.H_Boltzman.std():.4f}")
plt.xlabel("Tiempo")
plt.ylabel("H de Boltzmann")
plt.title("H de Boltzmann")
plt.grid(True)
plt.minorticks_on()
#plt.legend(loc='best', framealpha=0.3)
plt.tight_layout()

plt.savefig("H_Boltzman.png") 


# Sobre la energía y la presión

In [ ]:
df=pd.read_csv("gname2.txt", delim_whitespace=True, header=None, names=["t","Ep","Ec","Et"])

E_total=df.Et
Ec=df.Ec
Ep=df.Ep


plt.plot(t,Ep,"r-", label="Energía potencial")
plt.plot(t,Ec,"b-", label="Energía cinética")
plt.plot(t,E_total,"g-", label="Energía total")

plt.xlabel("Tiempo")
plt.ylabel("Energía")
plt.title("Energías")
plt.grid(True)
plt.minorticks_on()
plt.legend(loc='best', framealpha=0.3)
plt.tight_layout()
plt.savefig("Energias.png")

In [ ]:
plt.plot(t,E_total,"g-", label="Energía total")
plt.xlabel("Tiempo")
plt.ylabel("Energía")
plt.title("Energías")
plt.grid(True)
plt.minorticks_on()
plt.legend(loc='best', framealpha=0.3)
plt.tight_layout()
plt.savefig("Energia_total.png")



In [ ]:
print(E_total.agg(["mean","std"]))